cowprotocol/services
# Convenient way to measure future execution times

# Description
I was looking into instrumenting the execution times for Enso vs Tenderly gas estimates (my worry is Enso may be taking way longer) in a way that wouldn't add a ton of boilerplate code to our observe module (as I'm not sure it belongs there).

Inspired by how futures can be [instrumented](https://dtantsur.github.io/rust-openstack/src/tracing/instrument.rs.html#13-77), I created a `Future` extension which start a timer when first polled. The timer will automatically record its execution time when the future is dropped.

# Changes


- New `future` module in the `observe` crate, which implements a `Measure` type wrapping the inner future
- Upon first poll a timer is started. This will lead to recording a time once the future is dropped (if we want this to work with Shared futures we could also inspect the Output and report once it's ready)
- Use the new type in the Simulator to measure enso vs. tenderly simulation times.

## How to test
Run the system with either enso or tenderly simulator enabled e2e and place an order. Observe the following logs on http://localhost:11088/metrics

```
# HELP driver_future_execution_times Timing of measured futures.
# TYPE driver_future_execution_times histogram
driver_future_execution_times_bucket{label="enso_simulate_gas",le="0.005"} 0
driver_future_execution_times_bucket{label="enso_simulate_gas",le="0.01"} 0
driver_future_execution_times_bucket{label="enso_simulate_gas",le="0.025"} 0
driver_future_execution_times_bucket{label="enso_simulate_gas",le="0.05"} 0
driver_future_execution_times_bucket{label="enso_simulate_gas",le="0.1"} 0
driver_future_execution_times_bucket{label="enso_simulate_gas",le="0.25"} 72
driver_future_execution_times_bucket{label="enso_simulate_gas",le="0.5"} 73
driver_future_execution_times_bucket{label="enso_simulate_gas",le="1"} 75
driver_future_execution_times_bucket{label="enso_simulate_gas",le="2.5"} 75
driver_future_execution_times_bucket{label="enso_simulate_gas",le="5"} 75
driver_future_execution_times_bucket{label="enso_simulate_gas",le="10"} 75
driver_future_execution_times_bucket{label="enso_simulate_gas",le="+Inf"} 75
driver_future_execution_times_sum{label="enso_simulate_gas"} 11.483358082999995
driver_future_execution_times_count{label="enso_simulate_gas"} 75
```

## Related Issues

#1908
